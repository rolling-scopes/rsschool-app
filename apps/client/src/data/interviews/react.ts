import { InterviewTemplate } from './types';

export const reactTemplate: InterviewTemplate = {
  name: 'React interview ',
  examplesUrl: 'https://github.com/rolling-scopes-school/tasks/blob/master/react/questions.md',
  categories: [
    {
      id: 1000,
      name: 'JSX',
      questions: [
        { id: 1001, name: 'What is JSX?' },
        { id: 1002, name: 'Is it possible to use React without JSX?' },
      ],
    },
    {
      id: 1010,
      name: 'React DOM',
      questions: [
        { id: 1011, name: 'What is the virtual DOM? How does react use the virtual DOM to render the UI?' },
        { id: 1012, name: 'Is the virtual DOM the same as the shadow DOM?' },
        { id: 1013, name: 'What is the difference between the virtual DOM and the real DOM?' },
      ],
    },
    {
      id: 1020,
      name: 'Render',
      questions: [
        { id: 1021, name: 'When is a component rendered?' },
        { id: 1022, name: 'How not to render on props change?' },
        { id: 1023, name: 'Is it OK to use arrow functions in render methods?' },
      ],
    },
    {
      id: 1030,
      name: 'Interaction between components',
      questions: [
        { id: 1031, name: 'How do you pass a value from parent to child?' },
        { id: 1032, name: 'How do you pass a value from child to parent?' },
        { id: 1033, name: 'What is prop drilling?' },
        { id: 1034, name: 'Can a child component modify its own props?' },
        { id: 1035, name: 'How do you pass a value from sibling to sibling?' },
      ],
    },
    {
      id: 1040,
      name: 'Lifecycle and State',
      questions: [
        { id: 1041, name: 'What is the difference between props and state?' },
        { id: 1042, name: 'How does state in a class component differ from state in a functional component?' },
        { id: 1043, name: 'What is the component lifecycle?' },
        { id: 1044, name: 'How do you update lifecycle in function components?' },
      ],
    },
    {
      id: 1050,
      name: 'Ref',
      questions: [
        { id: 1051, name: 'What is the difference between refs and state variables?' },
        { id: 1052, name: 'When is the best time to use refs?' },
        { id: 1053, name: 'What is the proper way to update a ref in a function component?' },
      ],
    },
    {
      id: 1060,
      name: 'Context',
      questions: [
        { id: 1061, name: 'What is the difference between the context API and prop drilling?' },
        { id: 1062, name: "When shouldn't you use the context API?" },
      ],
    },
    {
      id: 1070,
      name: 'Redux',
      questions: [
        { id: 1071, name: 'Enumerate base principles' },
        { id: 1072, name: 'What is the typical flow of data in a React + Redux app?' },
        { id: 1073, name: 'Benefits of Redux?' },
      ],
    },
    {
      id: 1080,
      name: 'Other',
      questions: [
        { id: 1081, name: 'Is it a good idea to use Math.random for keys?' },
        { id: 1082, name: 'What are the limitations of React?' },
        { id: 1083, name: 'What is a higher order component?' },
        { id: 1084, name: 'What are uncontrolled and controlled components?' },
        { id: 1085, name: 'React optimizations' },
      ],
    },
    {
      id: 1100,
      name: 'Coding task',
      questions: [{ id: 1101, name: 'Small react app: form, button, results list' }],
    },
  ],
};
