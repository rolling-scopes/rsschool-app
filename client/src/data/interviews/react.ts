import { InterviewTemplate } from './types';

export const reactTemplate: InterviewTemplate = {
  name: 'React interview',
  examplesUrl: 'https://github.com/rolling-scopes-school/tasks/blob/master/react/questions.md',
  categories: [
    {
      id: 1000,
      name: 'JSX',
      questions: [
        { id: 1001, name: 'What is JSX?' },
        { id: 1002, name: 'Is it possible to use React without JSX?' },
      ],
    },
    {
      id: 1010,
      name: 'Styled-components',
      questions: [{ id: 1011, name: 'Using styled-components' }],
    },
    {
      id: 1020,
      name: 'Components',
      questions: [
        { id: 1021, name: 'Component Definition (Class-based, Functional)' },
        { id: 1022, name: 'Components Composition' },
      ],
    },
    {
      id: 1030,
      name: 'React DOM',
      questions: [
        { id: 1031, name: 'What is the virtual DOM? How does react use the virtual DOM to render the UI?' },
        { id: 1032, name: 'Is the virtual DOM the same as the shadow DOM?' },
        { id: 1033, name: 'What is the difference between the virtual DOM and the real DOM?' },
      ],
    },
    {
      id: 1040,
      name: 'Render',
      questions: [
        { id: 1041, name: 'When is a component rendered?' },
        { id: 1042, name: 'How not to render on props change?' },
        { id: 1043, name: 'Is it OK to use arrow functions in render methods?' },
        { id: 1044, name: 'Components rendering system' },
      ],
    },
    {
      id: 1050,
      name: 'Interaction between components',
      questions: [
        { id: 1051, name: 'How do you pass a value from parent to child?' },
        { id: 1052, name: 'How do you pass a value from child to parent?' },
        { id: 1053, name: 'What is prop drilling?' },
        { id: 1054, name: 'Can a child component modify its own props?' },
        { id: 1055, name: 'How do you pass a value from sibling to sibling?' },
      ],
    },
    {
      id: 1060,
      name: 'Lifecycle and State',
      questions: [
        { id: 1061, name: 'What is the difference between props and state?' },
        { id: 1062, name: 'How does state in a class component differ from state in a functional component?' },
        { id: 1063, name: 'What is the component lifecycle?' },
        { id: 1064, name: 'How do you update lifecycle in function components?' },
        { id: 1065, name: 'Controlled/uncontrolled components' },
        { id: 1066, name: 'Stateful vs stateless components' },
      ],
    },
    {
      id: 1070,
      name: 'Ref',
      questions: [
        { id: 1071, name: 'What is the difference between refs and state variables?' },
        { id: 1072, name: 'When is the best time to use refs?' },
        { id: 1073, name: 'What is the proper way to update a ref in a function component?' },
      ],
    },
    {
      id: 1080,
      name: 'Context',
      questions: [
        { id: 1081, name: 'What is the difference between the context API and prop drilling?' },
        { id: 1082, name: "When shouldn't you use the context API?" },
      ],
    },
    {
      id: 1090,
      name: 'Redux',
      questions: [
        { id: 1091, name: 'Enumerate base principles' },
        { id: 1092, name: 'What is the typical flow of data in a React + Redux app?' },
        { id: 1093, name: 'Benefits of Redux?' },
        { id: 1094, name: 'Async Redux flow' },
      ],
    },
    {
      id: 1100,
      name: 'State Management by managers',
      questions: [
        { id: 1101, name: 'Mobx' },
        { id: 1102, name: 'Reflux' },
      ],
    },
    {
      id: 1110,
      name: 'Routing',
      questions: [
        { id: 1111, name: 'React Router' },
        { id: 1112, name: 'History' },
      ],
    },
    {
      id: 1120,
      name: 'Other',
      questions: [
        { id: 1121, name: 'Is it a good idea to use Math.random for keys?' },
        { id: 1122, name: 'What are the limitations of React?' },
        { id: 1123, name: 'What is a higher order component?' },
        { id: 1124, name: 'What are uncontrolled and controlled components?' },
        { id: 1125, name: 'React optimizations' },
      ],
    },
    {
      id: 1130,
      name: 'Coding task',
      questions: [{ id: 1131, name: 'Small react app: form, button, results list' }],
    },
    {
      id: 1140,
      name: 'Next.js',
      questions: [
        { id: 1141, name: 'What is Next.js, and how does it differ from traditional React applications?' },
        { id: 1142, name: 'Explain the concept of server-side rendering (SSR) in Next.js' },
        { id: 1143, name: 'How does Next.js handle client-side routing?' },
        { id: 1144, name: 'What are the benefits of using Next.js for building React applications?' },
        {
          id: 1145,
          name: 'What is the purpose of the getInitialProps function in Next.js? How is it different from using getStaticProps or getServerSideProps?',
        },
        { id: 1146, name: 'How does Next.js handle automatic code splitting, and why is it important?' },
        {
          id: 1147,
          name: 'What is the purpose of the Link component in Next.js, and how does it differ from traditional anchor (<a>) tags?',
        },
        { id: 1148, name: 'How can you configure routing in a Next.js application?' },
      ],
    },
    {
      id: 1150,
      name: 'Data Fetching',
      questions: [
        {
          id: 1151,
          name: 'Compare and contrast getStaticProps and getServerSideProps in terms of use cases and performance considerations',
        },
        {
          id: 1152,
          name: 'When would you choose to use getStaticPaths in Next.js, and how does it relate to dynamic routes?',
        },
      ],
    },
    {
      id: 1160,
      name: 'Advanced Topics',
      questions: [
        { id: 1161, name: 'Describe the purpose and use cases of API routes in Next.js' },
        { id: 1162, name: 'Explain the concept of middleware in Next.js and its role in the application lifecycle' },
        { id: 1163, name: 'How does Next.js handle authentication in applications?' },
      ],
    },
  ],
};
