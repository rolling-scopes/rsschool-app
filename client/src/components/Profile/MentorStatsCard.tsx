import * as React from 'react';
import isEqual from 'lodash/isEqual';
import { List, Typography, Button, Tag } from 'antd';
import CommonCard from './CommonCard';
import MentorStatsModal from './MentorStatsModal';
import { MentorStats, Student } from 'common/models/profile';
import { ConfigurableProfilePermissions } from 'common/models/profile';
import { ChangedPermissionsSettings } from 'pages/profile';
import { CheckboxChangeEvent } from 'antd/lib/checkbox';
import { TeamOutlined, FullscreenOutlined } from '@ant-design/icons';

const { Text } = Typography;

type Props = {
  data: MentorStats[];
  isEditingModeEnabled: boolean;
  permissionsSettings?: ConfigurableProfilePermissions;
  onPermissionsSettingsChange: (event: CheckboxChangeEvent, settings: ChangedPermissionsSettings) => void;
};

type State = {
  courseIndex: number;
  isMentorStatsModalVisible: boolean;
};

class MentorStatsCard extends React.Component<Props, State> {
  state = {
    courseIndex: 0,
    isMentorStatsModalVisible: false,
  };

  private filterPermissions = ({ isMentorStatsVisible }: Partial<ConfigurableProfilePermissions>) => ({
    isMentorStatsVisible,
  });

  private showMentorStatsModal = (courseIndex: number) => {
    this.setState({ courseIndex, isMentorStatsModalVisible: true });
  };

  private hideMentortStatsModal = () => {
    this.setState({ isMentorStatsModalVisible: false });
  };

  private countStudents = (data: MentorStats[]) =>
    data.reduce((acc: Student[], cur: MentorStats) => (cur?.students?.length ? acc.concat(cur.students) : acc), [])
      .length;

  shouldComponentUpdate = (nextProps: Props, nextState: State) =>
    !isEqual(
      nextProps.permissionsSettings?.isMentorStatsVisible,
      this.props.permissionsSettings?.isMentorStatsVisible,
    ) ||
    !isEqual(nextProps.isEditingModeEnabled, this.props.isEditingModeEnabled) ||
    !isEqual(nextState.isMentorStatsModalVisible, this.state.isMentorStatsModalVisible);

  render() {
    const { isEditingModeEnabled, permissionsSettings, onPermissionsSettingsChange } = this.props;
    const stats = this.props.data;
    const { courseIndex, isMentorStatsModalVisible } = this.state;

    return (
      <>
        <MentorStatsModal
          stats={stats[courseIndex]}
          isVisible={isMentorStatsModalVisible}
          onHide={this.hideMentortStatsModal}
        />
        <CommonCard
          title="Mentor Statistics"
          icon={<TeamOutlined />}
          content={
            <>
              <div>
                <p>
                  Mentored Students:{' '}
                  <Text style={{ fontSize: 18 }} strong>
                    {this.countStudents(stats)}
                  </Text>
                </p>
                <p>
                  Courses as Mentor:{' '}
                  <Text style={{ fontSize: 18 }} strong>
                    {stats.length}
                  </Text>
                </p>
              </div>
              <List
                itemLayout="horizontal"
                dataSource={stats}
                renderItem={({ courseName, courseLocationName, students }, idx) => (
                  <List.Item style={{ display: 'flex', justifyContent: 'space-between' }}>
                    <div style={{ flexGrow: 2 }}>
                      <p style={{ fontSize: idx === 0 ? 20 : undefined, marginBottom: 5 }}>
                        <Text strong>
                          {courseName}
                          {courseLocationName && ` / ${courseLocationName}`}
                        </Text>
                      </p>
                      {students ? (
                        idx === 0 && (
                          <List
                            itemLayout="horizontal"
                            dataSource={students}
                            renderItem={({ githubId, name, isExpelled, totalScore }) => (
                              <List.Item style={{ display: 'flex', justifyContent: 'space-between' }}>
                                <div key={`mentor-students-${githubId} ${courseName}`} style={{ width: '100%' }}>
                                  <p
                                    style={{
                                      display: 'flex',
                                      justifyContent: 'space-between',
                                      fontSize: 12,
                                      marginBottom: 5,
                                    }}
                                  >
                                    <a href={`/profile?githubId=${githubId}`}>{name}</a>{' '}
                                    {isExpelled ? <Tag color="red">expelled</Tag> : <Tag color="green">active</Tag>}
                                  </p>
                                  <p style={{ fontSize: 12, marginBottom: 0 }}>
                                    Score: <Text mark>{totalScore}</Text>
                                  </p>
                                </div>
                              </List.Item>
                            )}
                          />
                        )
                      ) : (
                        <p>Does not have students at this course yet</p>
                      )}
                    </div>
                    {students && (
                      <Button
                        style={{ marginLeft: 16 }}
                        type="dashed"
                        onClick={this.showMentorStatsModal.bind(null, idx)}
                      >
                        <FullscreenOutlined />
                      </Button>
                    )}
                  </List.Item>
                )}
              />
            </>
          }
          permissionsSettings={permissionsSettings ? this.filterPermissions(permissionsSettings) : undefined}
          isEditingModeEnabled={isEditingModeEnabled}
          onPermissionsSettingsChange={onPermissionsSettingsChange}
        />
      </>
    );
  }
}

export default MentorStatsCard;
