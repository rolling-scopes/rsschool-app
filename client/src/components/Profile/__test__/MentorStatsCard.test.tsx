import React from 'react';
import { render } from '@testing-library/react';
import { MentorStatsCard } from '../MentorStatsCard';
import * as reactUse from 'react-use';

describe('MentorStatsCard', () => {
  const mentorStats = [
    {
      courseName: 'rs-2020-q1',
      courseLocationName: 'Minsk',
    },
    {
      courseName: 'rs-2018-q1',
      courseLocationName: 'Minsk',
      students: [
        {
          githubId: 'alex',
          name: 'Alex Petrov',
          isExpelled: false,
          totalScore: 3453,
          repoUrl: 'https://github.com/rolling-scopes-school/alex-RS2018Q1',
        },
        {
          githubId: 'vasya',
          name: 'Vasiliy Alexandrov',
          isExpelled: true,
          totalScore: 120,
          repoUrl: 'https://github.com/rolling-scopes-school/vasya-RS2018Q1',
        },
      ],
    },
  ];

  it('should render correctly', () => {
    jest.spyOn(reactUse, 'useAsync').mockReturnValue({ value: null, loading: false });

    const { container } = render(<MentorStatsCard githubId="test" data={mentorStats} />);
    expect(container).toMatchSnapshot();
  });
});
