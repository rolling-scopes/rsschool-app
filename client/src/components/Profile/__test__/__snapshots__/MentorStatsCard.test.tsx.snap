// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MentorStatsCard should render correctly 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered css-dev-only-do-not-override-w8mnev"
  >
    <div
      class="ant-card-head"
    >
      <div
        class="ant-card-head-wrapper"
      >
        <div
          class="ant-card-head-title"
        >
          <h2
            class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line css-dev-only-do-not-override-w8mnev"
            style="font-size: 16px; margin-bottom: 0px; display: flex; align-items: center; justify-content: space-between;"
          >
            <span>
              <span
                aria-label="team"
                class="anticon anticon-team"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="team"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M824.2 699.9a301.55 301.55 0 00-86.4-60.4C783.1 602.8 812 546.8 812 484c0-110.8-92.4-201.7-203.2-200-109.1 1.7-197 90.6-197 200 0 62.8 29 118.8 74.2 155.5a300.95 300.95 0 00-86.4 60.4C345 754.6 314 826.8 312 903.8a8 8 0 008 8.2h56c4.3 0 7.9-3.4 8-7.7 1.9-58 25.4-112.3 66.7-153.5A226.62 226.62 0 01612 684c60.9 0 118.2 23.7 161.3 66.8C814.5 792 838 846.3 840 904.3c.1 4.3 3.7 7.7 8 7.7h56a8 8 0 008-8.2c-2-77-33-149.2-87.8-203.9zM612 612c-34.2 0-66.4-13.3-90.5-37.5a126.86 126.86 0 01-37.5-91.8c.3-32.8 13.4-64.5 36.3-88 24-24.6 56.1-38.3 90.4-38.7 33.9-.3 66.8 12.9 91 36.6 24.8 24.3 38.4 56.8 38.4 91.4 0 34.2-13.3 66.3-37.5 90.5A127.3 127.3 0 01612 612zM361.5 510.4c-.9-8.7-1.4-17.5-1.4-26.4 0-15.9 1.5-31.4 4.3-46.5.7-3.6-1.2-7.3-4.5-8.8-13.6-6.1-26.1-14.5-36.9-25.1a127.54 127.54 0 01-38.7-95.4c.9-32.1 13.8-62.6 36.3-85.6 24.7-25.3 57.9-39.1 93.2-38.7 31.9.3 62.7 12.6 86 34.4 7.9 7.4 14.7 15.6 20.4 24.4 2 3.1 5.9 4.4 9.3 3.2 17.6-6.1 36.2-10.4 55.3-12.4 5.6-.6 8.8-6.6 6.3-11.6-32.5-64.3-98.9-108.7-175.7-109.9-110.9-1.7-203.3 89.2-203.3 199.9 0 62.8 28.9 118.8 74.2 155.5-31.8 14.7-61.1 35-86.5 60.4-54.8 54.7-85.8 126.9-87.8 204a8 8 0 008 8.2h56.1c4.3 0 7.9-3.4 8-7.7 1.9-58 25.4-112.3 66.7-153.5 29.4-29.4 65.4-49.8 104.7-59.7 3.9-1 6.5-4.7 6-8.7z"
                  />
                </svg>
              </span>
               
              Mentor Statistics
            </span>
          </h2>
        </div>
      </div>
    </div>
    <div
      class="ant-card-body"
    >
      <div>
        <p>
          Mentored Students:
           
          <span
            class="ant-typography css-dev-only-do-not-override-w8mnev"
            style="font-size: 18px;"
          >
            <strong>
              2
            </strong>
          </span>
        </p>
        <p>
          Courses as Mentor:
           
          <span
            class="ant-typography css-dev-only-do-not-override-w8mnev"
            style="font-size: 18px;"
          >
            <strong>
              2
            </strong>
          </span>
        </p>
      </div>
      <div
        class="ant-list ant-list-split css-dev-only-do-not-override-w8mnev"
      >
        <div
          class="ant-spin-nested-loading css-dev-only-do-not-override-w8mnev"
        >
          <div
            class="ant-spin-container"
          >
            <ul
              class="ant-list-items"
            >
              <li
                class="ant-list-item"
                style="display: flex; justify-content: space-between;"
              >
                <div
                  style="flex-grow: 2;"
                >
                  <p
                    style="font-size: 20px; margin-bottom: 5px;"
                  >
                    <span
                      class="ant-typography css-dev-only-do-not-override-w8mnev"
                    >
                      <strong>
                        rs-2020-q1
                         / Minsk
                      </strong>
                    </span>
                  </p>
                  <p>
                    Does not have students at this course yet
                  </p>
                </div>
              </li>
              <li
                class="ant-list-item"
                style="display: flex; justify-content: space-between;"
              >
                <div
                  style="flex-grow: 2;"
                >
                  <p
                    style="margin-bottom: 5px;"
                  >
                    <span
                      class="ant-typography css-dev-only-do-not-override-w8mnev"
                    >
                      <strong>
                        rs-2018-q1
                         / Minsk
                      </strong>
                    </span>
                  </p>
                </div>
                <button
                  class="ant-btn css-dev-only-do-not-override-w8mnev ant-btn-dashed"
                  style="margin-left: 16px;"
                  type="button"
                >
                  <span
                    aria-label="fullscreen"
                    class="anticon anticon-fullscreen"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="fullscreen"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M290 236.4l43.9-43.9a8.01 8.01 0 00-4.7-13.6L169 160c-5.1-.6-9.5 3.7-8.9 8.9L179 329.1c.8 6.6 8.9 9.4 13.6 4.7l43.7-43.7L370 423.7c3.1 3.1 8.2 3.1 11.3 0l42.4-42.3c3.1-3.1 3.1-8.2 0-11.3L290 236.4zm352.7 187.3c3.1 3.1 8.2 3.1 11.3 0l133.7-133.6 43.7 43.7a8.01 8.01 0 0013.6-4.7L863.9 169c.6-5.1-3.7-9.5-8.9-8.9L694.8 179c-6.6.8-9.4 8.9-4.7 13.6l43.9 43.9L600.3 370a8.03 8.03 0 000 11.3l42.4 42.4zM845 694.9c-.8-6.6-8.9-9.4-13.6-4.7l-43.7 43.7L654 600.3a8.03 8.03 0 00-11.3 0l-42.4 42.3a8.03 8.03 0 000 11.3L734 787.6l-43.9 43.9a8.01 8.01 0 004.7 13.6L855 864c5.1.6 9.5-3.7 8.9-8.9L845 694.9zm-463.7-94.6a8.03 8.03 0 00-11.3 0L236.3 733.9l-43.7-43.7a8.01 8.01 0 00-13.6 4.7L160.1 855c-.6 5.1 3.7 9.5 8.9 8.9L329.2 845c6.6-.8 9.4-8.9 4.7-13.6L290 787.6 423.7 654c3.1-3.1 3.1-8.2 0-11.3l-42.4-42.4z"
                      />
                    </svg>
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
